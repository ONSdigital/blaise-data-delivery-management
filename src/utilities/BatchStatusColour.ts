export function getDDFileStatusStyle(status: string): string {
    switch (status) {
        case "inactive":
            return "dead";
        case "in_arc":
            return "success";

        case "errored":
            return "error";
        default:
            return "pending";
    }
}
