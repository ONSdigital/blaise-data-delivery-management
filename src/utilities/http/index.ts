export * from "./dataDelivery";
export * from "./dataDeliveryStatus";
