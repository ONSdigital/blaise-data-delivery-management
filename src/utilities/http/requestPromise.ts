import axios, { Method } from "axios";
import { JSONValue } from "../../../Interfaces";
import { DataDeliveryFile, DataDeliveryBatchData, DataDeliveryFileStatus } from "../../../Interfaces";

type PromiseResponse = [number, DataDeliveryFile | DataDeliveryFileStatus | DataDeliveryBatchData | string | { [key: string]: string } | null];

async function requestPromiseJson(method: Method, url: string, body: JSONValue | null = null): Promise<PromiseResponse> {
    try {
        const response = await axios({
            url: url,
            method: method,
            data: body,
            validateStatus: () => true,
        });

        const data = response.data;

        if (!data) {
            return [response.status, null];
        }
        return [response.status, data];
    } catch (error) {
        console.log(error);
        throw error;
    }
}

type PromiseResponseList = [boolean, DataDeliveryBatchData[] | DataDeliveryFileStatus[] | { [key: string]: string }];

async function requestPromiseJsonList(method: Method, url: string, body: JSONValue | null = null): Promise<PromiseResponseList> {
    try {
        const response = await axios({
            url: url,
            method: method,
            data: body,
            validateStatus: () => true,
        });

        const data = response.data;

        if (response.status === 200) {
            if (!Array.isArray(data)) {
                return [false, []];
            }
            return [true, data];
        } else if (response.status === 404) {
            return [true, data];
        } else {
            return [false, []];
        }
    } catch (error) {
        console.log(error);
        throw error;
    }
}

export { requestPromiseJson, requestPromiseJsonList };