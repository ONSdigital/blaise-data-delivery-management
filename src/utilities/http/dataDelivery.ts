import {requestPromiseJson} from "./requestPromise";
type getBatchListResponse = [boolean, any[]];

function sendDataDeliveryRequest(): Promise<boolean> {
    console.log("Sending request to trigger data delivery");
    const url = "/api/trigger";

    return new Promise((resolve: (object: boolean) => void) => {
        requestPromiseJson("POST", url).then(([status, data]) => {
            console.log(`Response from trigger data delivery: Status ${status}, data ${data}`);
            if (status === 200) {
                if (data === "completed") {
                    resolve(true);
                    return;
                }
                resolve(false);
            } else {
                resolve(false);
            }
        }).catch((error: Error) => {
            console.error(`Failed to trigger data delivery, Error ${error}`);
            resolve(false);
        });
    });
}


export {sendDataDeliveryRequest};
