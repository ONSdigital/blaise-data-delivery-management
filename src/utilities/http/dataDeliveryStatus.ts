import { requestPromiseJson, requestPromiseJsonList } from "./requestPromise";
import { DataDeliveryBatchData, DataDeliveryFileStatus } from "../../../interfaces";

type getBatchInfoListResponse = [boolean, DataDeliveryFileStatus[]];
type getAllBatchesResponse = [boolean, DataDeliveryBatchData[]];
type getStatusDescriptionsResponse = [boolean, { [key: string]: string }];

async function getAllBatches(): Promise<getAllBatchesResponse> {
    console.log("Call to getAllBatches");
    const url = "/api/batch";

    try {
        const [success, data]: getAllBatchesResponse = await requestPromiseJsonList("GET", url);
        console.log(`Response from get Batch Info ${(success ? "successful" : "failed")}, data list length ${data.length}`);
        return [success, data];
    } catch (error) {
        console.error(`Response from get All Batches Failed: Error ${error}`);
        return [false, []];
    }
}

async function getBatchInfo(batchName: string): Promise<getBatchInfoListResponse> {
    console.log("Call to getBatchInfo");
    const url = `/api/batch/${batchName}`;

    try {
        const [success, data]: getBatchInfoListResponse = await requestPromiseJsonList("GET", url);
        console.log(`Response from get Batch Info ${(success ? "successful" : "failed")}, data list length ${data.length}`);
        return [success, data];
    } catch (error) {
        console.error(`Response from get Batch Info Failed: Error ${error}`);
        return [false, []];
    }
}

async function getBatchStatusDescriptions(): Promise<getStatusDescriptionsResponse> {
    let list: { [key: string]: string } = {};
    console.log("Call to getBatchStatusDescriptions");
    const url = "/api/state/descriptions";

    try {
        const [status, data] = await requestPromiseJson("GET", url);
        console.log(`Response from get Batch Status Descriptions ${status}, data ${data}`);
        if (status === 200) {
            list = data;
            return [true, list];
        } else if (status === 404) {
            return [true, list];
        } 
        else {
            return [false, list];
        }
    } catch (error) {
        console.error(`Response from get Batch Status Descriptions: Error ${error}`);
        return [false, list];
    }
}

export { getAllBatches, getBatchInfo, getBatchStatusDescriptions };
