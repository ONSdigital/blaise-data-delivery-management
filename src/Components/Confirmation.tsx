import React, {ReactElement, useState} from "react";
import {Link, Redirect, useHistory} from "react-router-dom";
import {ONSButton, ONSPanel} from "blaise-design-system-react-components";
import {sendDataDeliveryRequest} from "../utilities/http";

function Confirmation(): ReactElement {
    const [formError, setFormError] = useState<string>("");
    const [redirect, setRedirect] = useState<boolean>(false);
    const [confirm, setConfirm] = useState<boolean | null>(null);
    const [loading, setLoading] = useState<boolean>(false);
    const [message, setMessage] = useState<string>("");
    const history = useHistory();

    async function confirmOption() {
        if (confirm === null) {
            setFormError("Select an answer");
            return;
        }
        if (!confirm) {
            history.push("/");
            return;
        }

        setLoading(true);
        setMessage("");

        const success = await sendDataDeliveryRequest();

        setLoading(false);

        if (!success) {
            setMessage("Failed to trigger Data Delivery.");
            setRedirect(true);
            return;
        }

        setMessage("Triggered Data Delivery successfully, It may take a few minutes for the run to appear in the table below.");
        setRedirect(true);
    }

    return (
        <>
            {
                redirect && <Redirect
                    to={{
                        pathname: "/",
                        state: {status: message}
                    }}/>
            }
            <p>
                <Link to={"/"}>Previous</Link>
            </p>
            <h1>
                Are you sure you want to trigger Data Delivery?
            </h1>

            {
                message !== "" &&
                <ONSPanel status={message.includes("success") ? "success" : "error"}>
                    <p>{message}</p>
                </ONSPanel>
            }

            <form className="u-mt-m">
                {
                    formError === "" ?
                        confirmDeleteRadios(setConfirm)
                        :
                        <ONSPanel status={"error"}>
                            <p className="panel__error">
                                <strong>{formError}</strong>
                            </p>
                            {confirmDeleteRadios(setConfirm)}
                        </ONSPanel>
                }

                <br/>
                <ONSButton
                    label={"Continue"}
                    primary={true}
                    loading={loading}
                    id="confirm-continue"
                    onClick={() => confirmOption()}/>
                {!loading &&
                <ONSButton
                    label={"Cancel"}
                    primary={false}
                    id="cancel-overwrite"
                    onClick={() => history.push("/")}/>
                }
            </form>
        </>
    );
}

function confirmDeleteRadios(setConfirm: (value: (((prevState: (boolean | null)) => (boolean | null)) | boolean | null)) => void) {
    return (
        <fieldset className="fieldset">
            <legend className="fieldset__legend">
            </legend>
            <div className="radios__items">

                <p className="radios__item">
                        <span className="radio">
                        <input
                            type="radio"
                            id="confirm-overwrite"
                            className="radio__input js-radio "
                            value="True"
                            name="confirm-delete"
                            aria-label="No"
                            onChange={() => setConfirm(true)}
                        />
                        <label className="radio__label " htmlFor="confirm-overwrite">
                            Yes, trigger Data Delivery
                        </label>
                    </span></p>
                <br/>
                <p className="radios__item">
                        <span className="radio">
                        <input
                            type="radio"
                            id="cancel-keep"
                            className="radio__input js-radio "
                            value="False"
                            name="confirm-delete"
                            aria-label="Yes"
                            onChange={() => setConfirm(false)}
                        />
                        <label className="radio__label " htmlFor="cancel-keep">
                            No, do not trigger Data Delivery
                        </label>
                    </span>
                </p>
            </div>
        </fieldset>
    );
}

export default Confirmation;
