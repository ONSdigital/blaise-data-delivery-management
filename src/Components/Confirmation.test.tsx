/**
 * @jest-environment jsdom
 */
import React from "react";
import "@testing-library/jest-dom";
import { Switch, Route, Router } from "react-router-dom";
import { render } from "@testing-library/react";
import { screen } from "@testing-library/dom";
import { createMemoryHistory } from "history";
import userEvent from "@testing-library/user-event";
import Confirmation from "./Confirmation";
import { sendDataDeliveryRequest } from "../utilities/http";

const mockRoute = "/trigger";

afterAll(() => {
    jest.clearAllMocks();
});

describe("Check Confirmation page snapshot:", () => {
    it("matches the snapshot", async () => {
        const history = createMemoryHistory({
            initialEntries: [mockRoute]
        });
        const wrapper = render(
            <Router history={history}>
                <Confirmation />
            </Router>
        );
        
        expect(wrapper).toMatchSnapshot();
    });
});

describe("Check form:", () => {
    it("'Yes, trigger Data delivery' is selectable and is the only one checked", async () => {
        const history = createMemoryHistory({
            initialEntries: [mockRoute]
        });
        render(
            <Router history={history}>
                <Switch>
                    <Route path={mockRoute}>
                        <Confirmation />
                    </Route>
                </Switch>
            </Router>
        );
        
        const radioBtnForYes = screen.getByText("Yes, trigger Data Delivery");
        userEvent.click(radioBtnForYes);

        expect(screen.getByLabelText("Yes, trigger Data Delivery")).toBeChecked();
        expect(screen.getByLabelText("No, do not trigger Data Delivery")).not.toBeChecked();
    });
    
    it("'No, do not trigger Data delivery' is selectable and is the only one checked", async () => {
        const history = createMemoryHistory({
            initialEntries: [mockRoute]
        });
        render(
            <Router history={history}>
                <Switch>
                    <Route path={mockRoute}>
                        <Confirmation />
                    </Route>
                </Switch>
            </Router>
        );
        
        const radioBtnForNo = screen.getByText("No, do not trigger Data Delivery");
        userEvent.click(radioBtnForNo);

        expect(screen.getByLabelText("Yes, trigger Data Delivery")).not.toBeChecked();
        expect(screen.getByLabelText("No, do not trigger Data Delivery")).toBeChecked();
    });

    // NOTE: This behaviour is tested in trigger_data_delivery (maybe mock sendDataDeliveryRequest()???)
    it.skip("calls sendDataDeliverRequest() when 'Yes' is the selected form option and confirmed", async () => {
        const history = createMemoryHistory({
            initialEntries: [mockRoute]
        });
        render(
            <Router history={history}>
                <Switch>
                    <Route path={mockRoute}>
                        <Confirmation />
                    </Route>
                </Switch>
            </Router>
        );
        
        const radioBtnForYes = screen.getByText("Yes, trigger Data Delivery");
        const confirmBtn = screen.getByRole("button", { name: "Continue" });
        userEvent.click(radioBtnForYes);
        userEvent.click(confirmBtn);
        
        expect(sendDataDeliveryRequest).toHaveBeenCalled();
    });

    it("navigates back to the homepage when 'No' is the selected form option and confirmed", async () => {
        const history = createMemoryHistory({
            initialEntries: [mockRoute]
        });
        render(
            <Router history={history}>
                <Switch>
                    <Route path={mockRoute}>
                        <Confirmation />
                    </Route>
                </Switch>
            </Router>
        );
        
        const radioBtnForNo = screen.getByText("No, do not trigger Data Delivery");
        const confirmBtn = screen.getByRole("button", { name: "Continue" });
        userEvent.click(radioBtnForNo);
        userEvent.click(confirmBtn);
        
        expect(history.location.pathname).toEqual("/");
    });

    it("navigates back to the homepage when the 'Cancel' button is clicked", async () => {
        const history = createMemoryHistory({
            initialEntries: [mockRoute]
        });
        render(
            <Router history={history}>
                <Switch>
                    <Route path={mockRoute}>
                        <Confirmation />
                    </Route>
                </Switch>
            </Router>
        );
        
        const confirmBtn = screen.getByRole("button", { name: "Cancel" });
        userEvent.click(confirmBtn);
        
        expect(history.location.pathname).toEqual("/");
    });
});