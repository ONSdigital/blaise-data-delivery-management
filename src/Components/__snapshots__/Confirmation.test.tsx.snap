// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check Confirmation page snapshot: matches the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        aria-label="Breadcrumb"
        class="breadcrumb"
      >
        <ol
          class="breadcrumb__items u-fs-s"
        >
          <li
            class="breadcrumb__item"
            id="breadcrumb-0"
          >
            <a
              class="breadcrumb__link"
              href="/"
            >
              Home
            </a>
            <svg
              class="svg-icon"
              focusable="false"
              viewBox="0 0 8 13"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.74,14.28l-.57-.56a.5.5,0,0,1,0-.71h0l5-5-5-5a.5.5,0,0,1,0-.71h0l.57-.56a.5.5,0,0,1,.71,0h0l5.93,5.93a.5.5,0,0,1,0,.7L6.45,14.28a.5.5,0,0,1-.71,0Z"
                transform="translate(-5.02 -1.59)"
              />
            </svg>
          </li>
        </ol>
      </nav>
      <main
        class="page__main u-mt-no"
        id="main-content"
      >
        <h1
          class="u-mb-l"
        >
          Are you sure you want to trigger Data Delivery?
        </h1>
        <form
          class="u-mt-m"
        >
          <fieldset
            class="fieldset"
          >
            <legend
              class="fieldset__legend"
            />
            <div
              class="radios__items"
            >
              <p
                class="radios__item"
              >
                <span
                  class="radio"
                >
                  <input
                    aria-label="No"
                    class="radio__input js-radio "
                    id="confirm-overwrite"
                    name="confirm-delete"
                    type="radio"
                    value="True"
                  />
                  <label
                    class="radio__label "
                    for="confirm-overwrite"
                  >
                    Yes, trigger Data Delivery
                  </label>
                </span>
              </p>
              <br />
              <p
                class="radios__item"
              >
                <span
                  class="radio"
                >
                  <input
                    aria-label="Yes"
                    class="radio__input js-radio "
                    id="cancel-keep"
                    name="confirm-delete"
                    type="radio"
                    value="False"
                  />
                  <label
                    class="radio__label "
                    for="cancel-keep"
                  >
                    No, do not trigger Data Delivery
                  </label>
                </span>
              </p>
            </div>
          </fieldset>
          <br />
          <button
            class="btn "
            data-testid="button"
            id="confirm-continue"
            type="button"
          >
            <span
              class="btn__inner"
            >
              Continue
            </span>
          </button>
          <button
            class="btn btn--secondary "
            data-testid="button"
            id="cancel-overwrite"
            type="button"
          >
            <span
              class="btn__inner"
            >
              Cancel
            </span>
          </button>
        </form>
      </main>
    </div>
  </body>,
  "container": <div>
    <nav
      aria-label="Breadcrumb"
      class="breadcrumb"
    >
      <ol
        class="breadcrumb__items u-fs-s"
      >
        <li
          class="breadcrumb__item"
          id="breadcrumb-0"
        >
          <a
            class="breadcrumb__link"
            href="/"
          >
            Home
          </a>
          <svg
            class="svg-icon"
            focusable="false"
            viewBox="0 0 8 13"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.74,14.28l-.57-.56a.5.5,0,0,1,0-.71h0l5-5-5-5a.5.5,0,0,1,0-.71h0l.57-.56a.5.5,0,0,1,.71,0h0l5.93,5.93a.5.5,0,0,1,0,.7L6.45,14.28a.5.5,0,0,1-.71,0Z"
              transform="translate(-5.02 -1.59)"
            />
          </svg>
        </li>
      </ol>
    </nav>
    <main
      class="page__main u-mt-no"
      id="main-content"
    >
      <h1
        class="u-mb-l"
      >
        Are you sure you want to trigger Data Delivery?
      </h1>
      <form
        class="u-mt-m"
      >
        <fieldset
          class="fieldset"
        >
          <legend
            class="fieldset__legend"
          />
          <div
            class="radios__items"
          >
            <p
              class="radios__item"
            >
              <span
                class="radio"
              >
                <input
                  aria-label="No"
                  class="radio__input js-radio "
                  id="confirm-overwrite"
                  name="confirm-delete"
                  type="radio"
                  value="True"
                />
                <label
                  class="radio__label "
                  for="confirm-overwrite"
                >
                  Yes, trigger Data Delivery
                </label>
              </span>
            </p>
            <br />
            <p
              class="radios__item"
            >
              <span
                class="radio"
              >
                <input
                  aria-label="Yes"
                  class="radio__input js-radio "
                  id="cancel-keep"
                  name="confirm-delete"
                  type="radio"
                  value="False"
                />
                <label
                  class="radio__label "
                  for="cancel-keep"
                >
                  No, do not trigger Data Delivery
                </label>
              </span>
            </p>
          </div>
        </fieldset>
        <br />
        <button
          class="btn "
          data-testid="button"
          id="confirm-continue"
          type="button"
        >
          <span
            class="btn__inner"
          >
            Continue
          </span>
        </button>
        <button
          class="btn btn--secondary "
          data-testid="button"
          id="cancel-overwrite"
          type="button"
        >
          <span
            class="btn__inner"
          >
            Cancel
          </span>
        </button>
      </form>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
